Fix/vlfeat compliance

griwodz opened the pull request:
<!-- Checklist before submission:

 - I have read the [contribution guidelines](../CONTRIBUTING.md).
 - I have updated the documentation, if applicable.
 - I have ensured that the change is tested somewhere.
 - I have followed the prevailing code style (for history readability and limit conflicts for maintainance).

-->
## Description

The proposed change adds a descriptor computation that is compliant with VLFeat. The original PopSift descriptor was not because it used the angle differently: while VLFeat spreads the angular part of the descriptor over up to 8 bins of the descriptor's histogram, PopSift would only spread it over two. Furthermore, descriptors are arranged differently into the histograms (according to VLFeat documentation, VLFeat does that also differently from Lowe's original code).

## Features list

For backward compatibility, the VLFeat descriptor is not the default descriptor. The following settings lead to results that seem to be identical to the VLFeat 0.9.20 C code and command line (the 0.9.21 command line does not work):

```
--desc-mode=vlfeat              : use the VLFeat descriptor extraction
--norm-mode=classic             : use L2 normalization, not RootSift
--norm-multi 9 --write-as-uchar : multiply descriptor values with 2^9 and cast to uchar
--write-with-ori                : output of descriptors in the same format as VLFeat command line (instead of OpenCV-compliant)
--initial-blur 0.5              : PopSift default
--sigma 1.6                     : PopSift default
--threshold 0                   : set the peak threshold to 0
--edge-threshold 10.0           : PopSift default
```
<!--
- [ ] Feature one. Fix #XXX
- [ ] Improve something else
- [ ] Connect to #3 (to declare link to issues without closing it when the PR is merged).
- [X] Add "X" when it is done.
-->


## Implementation remarks

The L2 normalization had a bug that was removed.

The orientation computation does now use a bilateral filter, which does not appear to have any visible effect. This is a change from the previous PopSift default, and it may be safer to make it a VLFeat-specific option.

The threshold for accepting an orientation in VLFeat is computed before smoothing the histogram, PopSift did it after computing the histogram (consequently accepting fewer orientations). This was change to work like VLFeat.

<!--
Explain main implementation choices.
It is also the right place to ask for feedback and help when you hesitate on the implementation.
-->

Commit:
[popsift] choosing a runtime between two descriptor printing mode

Commit:
[bugfix] vlfeat compliance in orientation computation

Commit:
Merge branch 'develop' into fix/misalignBugOri

Commit:
[demo] print default values in usage

Commit:
[bug] corrected truncating in L2 normalization

Commit:
[bug] added reverse square root required by previous bugfix

Commit:
[feature] VLFeat-compliant descriptor extraction

Commit:
[popsift] cleanup

Commit:
[popsift] first parallelism in vlfeat descriptors

Commit:
[bugfix] limit accepted orientations to histogram entries 0.8*max again

Commit:
[popsift] use 32 threads for each descriptor

Commit:
[popsift] commented extended option

griwodz commented on src/popsift/s_desc_norm_l2.h:
```diff
@@ -96,14 +96,14 @@ void NormalizeL2::normalize( const float* src_desc, float* dst_desc, const bool
     }
     norm = popsift::shuffle( norm,  0 );
 
-    descr.x = min( descr.x, 0.2f*norm );
```
I made a direct comparison between the develop branch and develop with the L2 normalization bug fixed.
I compare the descriptors extracted from all images in the dataset containing boat, bark, bikes, graf, leuven, tree, ubc and wall.
The average number of descriptors extracted per image was 10000.
The largest difference before and after the fix was 2 descriptors changed in 1 value.

simogasp commented on src/application/main.cpp:
```diff
@@ -64,13 +65,13 @@ static void parseargs(int argc, char** argv, popsift::Config& config, string& in
             ("octaves", value<int>(&config.octaves), "Number of octaves")
             ("levels", value<int>(&config.levels), "Number of levels per octave")
-            ("sigma", value<float>()->notifier([&](float f) { config.setSigma(f); }), "Initial sigma value")
+            ("sigma", value<float>()->notifier([&](float f) { config.setSigma(f); }), "Initial sigma value (default 1.6)")
```
```suggestion
            ("sigma", value<float>()->notifier([&](float f) { config.setSigma(f); })->default_value(config.getSigma()), "Initial sigma value")
```
using `->default_value()` will automatically print the value when calling `--help`. This is preferable than the raw value as then it becomes difficult to maintain if it ever changes etc.
The same for the others.

simogasp commented on src/application/main.cpp:
```diff
@@ -81,9 +82,9 @@ static void parseargs(int argc, char** argv, popsift::Config& config, string& in
+        "loop, iloop, grid, igrid, notile, vlfeat\n"
 	"Default is loop\n"
-        "loop is OpenCV-like horizontal scanning, computing only valid points, grid extracts only useful points but rounds them, iloop uses linear texture and rotated gradiant fetching. igrid is grid with linear interpolation. notile is like igrid but avoids redundant gradiant fetching.")
+        "loop is OpenCV-like horizontal scanning, computing only valid points, grid extracts only useful points but rounds them, iloop uses linear texture and rotated gradiant fetching. igrid is grid with linear interpolation. notile is like igrid but avoids redundant gradiant fetching, vlfeat replicates vlfeat behaviour.")
```
maybe briefly remind the vlfeat behaviour like done for opencv for example

griwodz replied:
I have extended it and move it to Config::getDescModeUsage(). The usage is getting very long, I wonder if we should re-write this to support --desc-mode=?

simogasp commented on src/popsift/sift_conf.h:
```diff
@@ -93,7 +93,9 @@ struct Config
         /// variant of IGrid, no duplicate gradient fetching
-        NoTile
+        NoTile,
+        /// extraction code according to VLFeat
```
same here, maybe briefly remind the difference so we have it in the doc

simogasp commented on src/popsift/s_orientation.cu:
```diff
@@ -55,6 +60,7 @@ inline float compute_angle( int bin, float hc, float hn, float hp )
 /*
  * Histogram smoothing helper
  */
+#ifdef WITH_VLFEAT_SMOOTHING
```
it would be better to use `POPSIFT_WITH_VLFEAT_SMOOTHING` or `POPSIFT_USE_VLFEAT_SMOOTHING` to avoid conflict and for consistency through repositories.

Also, is this supposed to be an option the user can choose or is it gonna be a kind of debugging or dev-only feature? 
In the first case, we should then also add the option in cmake and add the definition at compile time

```cmake
option(PopSift_USE_VLFEAT_SMOOTHING "Use VLFeat smoothing." ON)

...

# set the variable
if(PopSift_USE_VLFEAT_SMOOTHING)
  set(POPSIFT_USE_VLFEAT_SMOOTHING   1)
else()
  set(POPSIFT_USE_VLFEAT_SMOOTHING   0)
endif()

...

# add also the message in the final summary
message("******************************************")
message("Building configuration:\n")
message(STATUS "PopSift version: " ${PROJECT_VERSION})
message(STATUS "Build type: " ${CMAKE_BUILD_TYPE})
...
message(STATUS "Using VLFeat smoothing: " ${POPSIFT_USE_VLFEAT_SMOOTHING})
```

and add the entry in `cmake/sift_config.h`

```c++
#define POPSIFT_USE_VLFEAT_SMOOTHING()            @POPSIFT_USE_VLFEAT_SMOOTHING@
```

and `#include <popsift/sift_config.h>` wherever it is used

griwodz replied:
I think we can remove the alternative and keep only VLFEAT_SMOOTHING. This is from the original development and the other setting has never been used.

simogasp commented on src/popsift/s_orientation.cu:
```diff
@@ -28,6 +28,11 @@
+/* Bilinear histogram seems to be VLFeat's default now.
+ * We emulate it because we have different orientations without it.
+ */
+#define WITH_BILINEAR_ORI
```
same comment as `WITH_VLFEAT_SMOOTHING`

griwodz replied:
I'm just comparing the effect of every little change on the descriptors. I think that the difference is big enough to make WITH_BILINEAR_ORI a Config parameter instead of a define.

simogasp commented on src/popsift/s_orientation.cu:
```diff
@@ -177,46 +214,51 @@ void ori_par( const int           octave,
-        bool predicate = ( bin < ORI_NBINS ) && ( sm_hist[bin] > max( sm_hist[prev], sm_hist[next] ) );
+        bool predicate = ( bin < ORI_NBINS ) && ( sm_hist[bin] > max( sm_hist[prev], sm_hist[next] ) ) && (sm_hist[bin] > 0.8f * maxval);
 
+#define REFINE 2
```
maybe we can use `WITH_VLFEAT_SMOOTHING` or `WITH_BILINEAR_ORI`?

griwodz replied:
I have removed WITH_VLFEAT_SMOOTHING because we have used only "true" for years.
I have replaced WITH_BILINEAR_ORI with a command line switch --ori-mode=BestBin or InterpolatedBin. The first is original PopSift behaviour, the second current VLFeat behaviour. But note that VLFeat is also using a #define to support BestBin instead.

Commit:
[popsift] avoid a conflict between old and vlfeat orientation computation

Commit:
[cleanup] remove histogram smoothing mode that was never used

Commit:
[cleanup] remove unused code

Commit:
[popsift] move rotation compensation to descriptor extraction

Commit:
Merge branch 'develop' into fix/vlfeatCompliance

Commit:
[popsift] choose between orientation computation modes

BestBin is the original PopSift approach. InterpolatedBin is a mode that emulates VLFeat behaviour.

Commit:
[bugfix] fix half-bin shift in Interpolated mode

Commit:
[popsift] minor range modification in orientation computation

Commit:
[doc] improved doc for enum DescMode

Commit:
[doc] moved the Usage for the DescMode parameter to sift_conf.cu

griwodz committed:
[doc] Integrating Config defaults into the Usage.

[doc] Integrating Config defaults into the Usage.

Co-authored-by: Simone Gasparini <simone.gasparini@gmail.com>

griwodz commented:
I have a few small illustrations for a test between VLFeat and PopSift with the appropriate switches. These illustrations are for boat/img1.pgm from the Oxford dataset.

The first picture shows the L2 distances in pixels that have been found by brute-force matching between PopSift and VLFeat descriptors (most distances are 0 or .9e-5, so I had to add 1 for the logscale image).
![pixdist](https://user-images.githubusercontent.com/2108546/95476148-3804b680-0987-11eb-8238-46bf81d61f3f.png)

The second picture shows the matching points, sorted by the L2 distances between their descriptors in purple, and the L2 distance between the PopSift descriptor and the 2nd best matching descriptor from VLeat.
![descriptordist-1st-2nd](https://user-images.githubusercontent.com/2108546/95476801-ed376e80-0987-11eb-918a-05e659fe05bd.png)

The third picture shows the same data, but this time sorted by the distances to the 2nd best match.
![descriptordist-2nd-1st](https://user-images.githubusercontent.com/2108546/95477126-4bfce800-0988-11eb-8dbe-55c166c4ca2b.png)

So there will be false matches.

griwodz commented:
And now I'm showing the same figures WITHOUT the VLFeat-specific switches, but with some tiny changes that were side-effects of trying to get closer to VLFeat.

L2-distance between the pixels of PopSift default mode and VLFeat:
![v2-pixdist](https://user-images.githubusercontent.com/2108546/95480994-cc254c80-098c-11eb-8c07-88502fd95622.png)

L2-distance between descriptors of PopSift default mode and VLFeat, best matches in purple, 2nd best matches in green:
![v2-2](https://user-images.githubusercontent.com/2108546/95481083-e65f2a80-098c-11eb-9d3a-fdd7975c46c2.png)

griwodz commented:
Seeing the descriptor distances, I have to go back to the code and check if there's a place in the code where I'm swapping the flags, because this is weird.

Commit:
[popsift] added Config::getSigma()

griwodz commented:
@fabiencastan So now I have compared descriptor distances between the develop branch, the new "VLFeat" mode and the develop branch with a few tiny changes to be more similar to VLFeat. When I look at the direct comparison, I get very confused why voctree is failing with the old code. Look at the two figures.

![old-vs-new-part1](https://user-images.githubusercontent.com/2108546/95501260-4ca57680-09a8-11eb-81bf-805aa3dd0122.png)

![old-vs-new-page2](https://user-images.githubusercontent.com/2108546/95501282-5333ee00-09a8-11eb-9e62-fbb79e822f88.png)

Commit:
[popsift] rotate orientation by 180deg in all older modes

Commit:
[test] draft comparison code

Commit:
[popsift] remove circular limitation for descriptor in VLFeat mode

griwodz commented:
@simogasp @fabiencastan Now it is really time to compare the VLFeat-compatible descriptor with the VLFeat-trained voctree. Our original orientation computation was never the problem, the descriptor computation is different. This led to a different probability distribution of values in the descriptors' 128 floats, which explains why the voctree code couldn't find matches with PopSift.

griwodz commented:
This is the difference in the individual cells of the old PopSift descriptor. The 128 values are 4x4 histograms of 8 rotational bins each. The figure shows the 8 bins sides by side, for each bin the **average** difference between VLFeat and PopSift descritors for a 4x4 group.

![descdist-boat1-new](https://user-images.githubusercontent.com/2108546/95686715-08141800-0c00-11eb-858b-f0b5fb098d08.png)

You can see that the PopSift descriptor values are generally higher in the first bin, lower in the later bins.

griwodz commented:
Now, with the VLFeat-compatible descriptor, it looks like this:

![descdist-boat1-vlfeatdesc](https://user-images.githubusercontent.com/2108546/95686763-57f2df00-0c00-11eb-8327-f7d35554e032.png)

griwodz commented:
The success for correct matching is also quite a bit better. The figure shows L2 distance between the best descriptors (in purple) between the old descriptor on the left and VLFeat-compatible descriptor in the right.

![descriptor](https://user-images.githubusercontent.com/2108546/95687041-8624ee80-0c01-11eb-9c3b-f34db5f064f1.png)

Commit:
[popsift] avoid boost in libpopsift

Commit:
[test] updating descriptor file comparison

Commit:
[testing] remove test file, prioritize main PR

simogasp commented on src/popsift/sift_conf.cu:
```diff
@@ -8,12 +8,25 @@
+{
+    std::for_each( l.begin(), l.end(), [](char& c) { c = ::tolower(c); });
+    std::for_each( r.begin(), r.end(), [](char& c) { c = ::tolower(c); });
+    if( !strcmp( l.c_str(), r.c_str() ) )
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
```
```suggestion
    return l == r;
```

simogasp commented on src/popsift/sift_conf.cu:
```diff
@@ -8,12 +8,25 @@
 
+static bool stringIsame( string l, string r )
+{
+    std::for_each( l.begin(), l.end(), [](char& c) { c = ::tolower(c); });
```
```suggestion
    // early exit
    if(l.size() != r.size())
    {
        return false;
    }

    std::for_each( l.begin(), l.end(), [](char& c) { c = ::tolower(c); });
```

griwodz replied:
I take the proposed change. I wanted to go for strncmp, but Windows does apparently not have it, it has stricmp. And adding Boost just for this seemed like too much.

griwodz replied:
@simogasp It would be really preferable to use strcasecmp() for non-Windows and _stricmp() for Windows. The first requires inclusion of strings.h (at least on Linux and Mac) and the other string.h. The first one is a function, the second is a macro. I'm too stupid to write a CMake test, so I gave up and wrote these crappy lines to make it run.
And is speed essential? It should be used to evaluate the command line.

simogasp replied:
I think it is ok like this, it is just for comparing the string in the enum. And I agree that it's not worth to re-add boost just for this.

griwodz committed:
[popsift] shorter replace strcmp with string::op==

Co-authored-by: Simone Gasparini <simone.gasparini@gmail.com>

Commit:
[popsift] add more Config::get* functions, improve usage

griwodz commented:
Btw.: the nice results were created with the following command line:
`popsift-demo --pgmread-loading --norm-mode=classic --norm-multi 9 --write-as-uchar --write-with-ori --initial-blur 0.5 --sigma 1.6 --threshold 0 --edge-threshold 10.0 --desc-mode=vlfeat -i boat.pgm`

Commit:
[popsift] remove failed interpolating orientation code
